engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java

```diff
@@ -0,0 +1,102 @@
+package org.enso.interpreter.node.expression.builtin.mutable;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.TruffleLanguage.ContextReference;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.CachedContext;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import java.util.Arrays;
+import java.util.Comparator;
+import org.enso.interpreter.Language;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.node.callable.InvokeCallableNode;
+import org.enso.interpreter.node.callable.InvokeCallableNode.ArgumentsExecutionMode;
+import org.enso.interpreter.node.callable.InvokeCallableNode.DefaultsExecutionMode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.callable.argument.CallArgumentInfo;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.Stateful;
+import org.enso.interpreter.runtime.state.data.EmptyMap;
+import org.enso.interpreter.runtime.type.TypesGen;
+
+@BuiltinMethod(type = "Array", name = "sort", description = "Sorts a mutable array in place.")
+public abstract class SortNode extends Node {
+  private final ConditionProfile consProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile lessProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile equalsProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile greaterProfile = ConditionProfile.createCountingProfile();
+
+  abstract Object execute(VirtualFrame frame, Object _this, Object comparator);
+
+  static SortNode build() {
+    return SortNodeGen.create();
+  }
+
+  @Specialization
+  Object doSort(
+      VirtualFrame frame,
+      Array _this,
+      Object comparator,
+      @Cached("buildInvokeNode()") InvokeCallableNode invokeNode,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef,
+      @Cached("ctxRef.get().getBuiltins().ordering().newLess()") Atom less,
+      @Cached("ctxRef.get().getBuiltins().ordering().newEqual()") Atom equal,
+      @Cached("ctxRef.get().getBuiltins().ordering().newGreater()") Atom greater) {
+    Comparator<Object> compare =
+        (l, r) -> {
+          Stateful result =
+              invokeNode.execute(comparator, frame, EmptyMap.create(), new Object[] {l, r});
```

Comment:
I'm surprised this works at all. It captures a virtual frame reference and then leaks it to the other side of a truffle boundary. Please move the construction of this comparator to a separate static method (so that we fully control what is captured in the lambda scope) and make it take a `MaterializedFrame` in the argument.